src/main/java/fi/aalto/cs/apluscourses/ui/exercise/ExercisesView.java

```diff
@@ -32,12 +58,26 @@ public JPanel getBasePanel() {
    * Sets the view model of this view, or does nothing if the given view model is null.
    */
   public void viewModelChanged(@Nullable ExercisesTreeViewModel viewModel) {
-    ApplicationManager.getApplication().invokeLater(
-        () -> exerciseGroupsTree.setViewModel(viewModel),
+    ApplicationManager.getApplication().invokeLater(() -> {
+      exerciseGroupsTree.setViewModel(viewModel);
+      hasCourse = mainViewModel.getHasCourse();
+      updateComponents();
+    },
         ModalityState.any()
     );
   }
 
+  private void updateComponents() {
+    emptyText.setText(getText("ui.module.ModuleListView.turnIntoAPlusProject"));
+    emptyText.setHorizontalAlignment(SwingConstants.CENTER);
+    emptyText.setVerticalAlignment(SwingConstants.CENTER);
+    if (hasCourse) {
+      cl.show(cardPanel,"TreeCard");
+    } else {
+      cl.show(cardPanel, "LabelCard");
+    }
```

Comment:
This logic could be moved to view model (and unit tested) if there was a method in the view model which returns the name of the card to be shown. Then here it would simply be:
```java
cl.show(cardPanel, mainViewModel.getCardToBeShown());
```

Reply:
The following is by no means required, just an idea:

Should you like, you could write `CardPanel` class that inherits `JPanel` and has `getCurrentCardName` and `setCurrentCardName` methods. Then you could add `currentCardNameBindable` field to that class and bind it to the corresponding observable property in the view model. And tada: whenever that property changes, the correct card is automatically shown without any "update view" method calls.

If you want to try this, I'd be glad to give further help.